src/couch_epi_server.erl

```diff
@@ -0,0 +1,138 @@
+
+%% ------------------------------------------------------------------
+%% gen_server Function Exports
+%% ------------------------------------------------------------------
+
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-record(epi_server_state, {subscriptions}).
+
+%% ------------------------------------------------------------------
+%% API Function Definitions
+%% ------------------------------------------------------------------
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+subscribe(App, Key, MFA) ->
+    subscribe(?SERVER, App, Key, MFA).
+
+subscribe(Server, App, Key, {_M, _F, _A} = MFA) ->
+    gen_server:call(Server, {subscribe, App, Key, MFA}).
+
+unsubscribe(Subscription) ->
+    unsubscribe(?SERVER, Subscription).
+
+unsubscribe(Server, Subscription) ->
+    gen_server:call(Server, {unsubscribe, Subscription}).
+
+notify(App, Key, OldData, Data) ->
+    notify(?SERVER, App, Key, OldData, Data).
+
+notify(Server, App, Key, OldData, Data) ->
+    gen_server:cast(Server, {notify, App, Key, OldData, Data}).
+
+
+%% ------------------------------------------------------------------
+%% gen_server Function Definitions
+%% ------------------------------------------------------------------
+
+init(_Args) ->
+    State = #epi_server_state{subscriptions = dict:new()},
+    {ok, State}.
+
+handle_call({subscribe, App, Key, MFA}, _From,
+        #epi_server_state{subscriptions = Subscriptions0} = State0) ->
+    {Subscription, Subscriptions1} = add(Subscriptions0, App, Key, MFA),
+    State1 = State0#epi_server_state{subscriptions = Subscriptions1},
+    {reply, {ok, Subscription}, State1};
+handle_call({unsubscribe, Subscription}, _From,
+        #epi_server_state{subscriptions = Subscriptions0} = State0) ->
+    Subscriptions1 = remove(Subscriptions0, Subscription),
+    State1 = State0#epi_server_state{subscriptions = Subscriptions1},
+    {reply, ok, State1};
+handle_call(_Request, _From, State) ->
+    {stop, normal, State}.
+
+handle_cast({notify, App, Key, OldData, Data},
+        #epi_server_state{subscriptions = Subscriptions} = State) ->
+    Subscribers = subscribers(Subscriptions, App, Key),
+    notify_subscribers(Subscribers, App, Key, OldData, Data),
+    {noreply, State};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+subscribers(Subscriptions, App, Key) ->
+    case dict:find({App, Key}, Subscriptions) of
+        error ->
+            [];
+        {ok, Subscribers} ->
+            Subscribers
+    end.
+
+add(Subscriptions, App, Key, MFA) ->
+    Subscription = make_ref(),
```

Comment:
Rather than make_ref, you should pass the From through and use a monitor so that you can remove the subscription if the subscriber.